Added a _getting started_ guide to set up a blank environment that can be used to deploy services. (#50)
stackabletech/documentation: README.adoc

=== Overview
We will perform the following steps in order to arrive at a working platform, the don't absolutely have to be done in the order listed here, but I'd suggest that you know what you are doing and why you are doing it if you want to deviate.

. Install Kubernetes
. Add Nodes to Kubernetes
. Install operators for products

While you can run the entire platform on your laptop, or a single virtual machine, but it is highly recommended to use several machines for the following steps.
The following terms will be used throughout the remainder of this guide:

* *Orchestrator*: A single server which is used to run components that are needed only once
* *Worker*: One or more servers that are intended to run the actual workloads (HDFS, Kafka, ...) later on

NOTE: Going forward, we will assume that you are installing this on several machines, but most commands should work the same in a single-machine scenario.
We will alert you if something needs to be done differently when running everything on the same machine.

.By the end of this guide, the environment that you have set up will look like this

> image::images/quickstart_architecture.png[]

Reviewer: Image is not part of the PR - please add.

Reviewer 2: I'll review this by using the guide to work on Issue https://github.com/stackabletech/t2/issues/11

    apt-get install gnupg

The key is available on the Ubuntu keyserver under the id _16dd12f5c7a6d76a_ and can be added to your apt keychain with the following command:

    apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 16dd12f5c7a6d76a

To enable the actual repository, put the following content into `/etc/apt/sources.list.d/stackable.list` :

    deb https://repo.stackable.tech/repository/deb-nightly buster main

===== Centos/RHEL

*coming soon*

=== Install Kubernetes
Usually the installation and configuration of Kubernetes is a full-blown project all of its own, so it might seem a bit weird to find this tucked away in a little section of the quickstart guide like this.
However, the Stackable Platform only uses a very limited subset of the full Kubernetes features, so for our purposes we could actually get away with just running a https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/[kube-apiserver] and leaving out the entire rest of the https://kubernetes.io/docs/concepts/overview/components/[control-plane].
This makes the deployment a lot simpler.

> That being said, Stackable will work with any Kubernetes, please feel free to skip this chapter if you already have a working Kubernetes cluster, or want to use a managed Kubernetes offering like the ones provided by https://www.ionos.com/enterprise-cloud/managed-kubernetes[IONOS], https://cloud.google.com/kubernetes-engine[Google] or https://docs.microsoft.com/en-us/azure/aks/[Azure].

Reviewer 3: Are we sure they allow custom nodes?
(IONOS do but the others?)
Maybe add a note that we haven't tested them.

> If you choose to follow one of the options mentioned above, feel free to skip ahead to https://xxx[xxx], as you won't need the information in the rest of this chapter.

Reviewer 3: ....

> Since you are still reading this we'll assume that you don't have easy access to an existing Kubernetes and instead install one.

Reviewer 3: This reads weird "instead install one". Maybe we can just remove this sentence?

There are numerous possibilities to go about this, but we will focus on https://k3s.io/[K3S], a lightweight Kubernetes distribution that is provided by https://rancher.com/[Rancher].

> Run the following command on your _orchestrator_ server:

Reviewer 3: You used `node` above instead of `server`

>     curl -sfL https://get.k3s.io | sh -

Reviewer 3: Maybe mention something about versions?
We only tested Kubernetes 1.18 and above.

This will download and install K3S, and configure systemd services to run at startup.
Client tools like `kubectl` will also be installed and configured correctly to connect to Kubernetes.

To check if everything worked as expected you can use kubctl to retrieve the cluster information:

    kubectl cluster-info

See below for an example of an expected result:

----
Kubernetes control plane is running at https://127.0.0.1:6443
CoreDNS is running at https://127.0.0.1:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy
Metrics-server is running at https://127.0.0.1:6443/api/v1/namespaces/kube-system/services/https:metrics-server:/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
----

> ==== Distribute KubeConfig

Reviewer 3: You use KubeConfig and Kubeconfig. Not sure which is correct but it'd be best to be consistent.

NOTE: Currently the KubeConfig will contain 127.0.0.1 as the api-server IP, you will need to change this to the ip address of your orchestrator server

This file needs to be put on every node of the cluster - ideally you'd copy this file to `/root/.kube/config` which will make all components that run as the root user use this configuration.
Perform this step for the orchestrator also, as the operators will expect to finde a KubeConfig in that place.

If you need to be able to connect to other Kubernetes clusters, or do not want to make this the global config (for the root user) for other reasons, then you can choose a different location for this file.

Please see below for extra steps that need to be taken to enable this config.

=== Specify a Repository
Stackable downloads packages from repositories, which the agents need to know about.
To avoid having to configure these repositories on every agent they are instead specified in Kubernetes and retrieved from there by the agent.

In order to allow creating a repository, you'll have to create the CRD for repositories in your freshly installed Kubernetes cluster.
The CRD looks like this:

> 

Reviewer 3: ```suggestion

[source,yaml]
```

----
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:

>   name: repositories.stable.stackable.de

Reviewer 3: We should change this to stackable.tech (also in the group below)
This'll probably need a follow-up issue in the agent?

Author: I'll need to check in the agent, but most probably, yes. Can we leave as is for now to get this merged?

spec:
  group: stable.stackable.de
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                repo_type:
                  type: string
                properties:
                  type: object
                  additionalProperties:
                    type: string
  scope: Namespaced
  names:
    plural: repositories
    singular: repository
    kind: Repository
    shortNames:
    - repo
----

You can choose whatever way is most convenient for you to apply this CRD to your cluster, one possible way is using `kubectl` from the command line:

> 

Reviewer 3: ```suggestion

[source,bash]
```

Not sure if that works

----
cat <<EOF | kubectl apply -f -
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: repositories.stable.stackable.de
spec:
  group: stable.stackable.de
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                repo_type:
                  type: string
                properties:
                  type: object
                  additionalProperties:
                    type: string
  scope: Namespaced
  names:
    plural: repositories
    singular: repository
    kind: Repository
    shortNames:
    - repo
EOF
----

> You can either host your own repository, if you so choose, or specify our public repository for convenience.

Reviewer 4: You can either host your own repository or specify our public repository for convenience.

The specification for our repository is shown below and can be applied with `kubectl` just like the definition above:

> 

Reviewer 3: ```suggestion

[source,bash]
```
see above

----
cat <<EOF | kubectl apply -f -
apiVersion: "stable.stackable.de/v1"
kind: Repository
metadata:
  name: stackablepublic
spec:
  repo_type: StackableRepo
  properties:
    url: https://repo.stackable.tech/repository/packages/
EOF
----

=== Install Operators

> Depending on which products you want to manage with the Stackable platform you will want to install at least one of our operators.

Reviewer 4: Depending on the products you want to manage with the Stackable platform, you need to install our corresponding operators.

They are made available as OS packages as well, the currently available operators are:

|===
|Product |Package Name

|Apache Kafka
|stackable-kafka-operator-server

|Apache Spark
|stackable-spark-operator-server

|Apache Zookeeper
|stackable-zookeeper-operator-server
|===

> 

Reviewer: I'd add an example here - as done below for the agent or add a link to the respective project in github

===== RHEL

    *coming soon*

===== Ubuntu

    apt-get install stackable-spark-operator-server

Reviewer 5: 👍

To start and enable the services please run the following commands:

----
# Kafka
systemctl start stackable-kafka-operator-server
systemctl enable stackable-kafka-operator-server

# Spark
systemctl start stackable-spark-operator-server
systemctl enable stackable-spark-operator-server

# Zookeeper
systemctl start stackable-zookeeper-operator-server
systemctl enable stackable-zookeeper-operator-server
----

To check if the services started succesfully you can run (example for Zookeeper):

----
systemctl status stackable-zookeeper-operator-server
----

This should report the service as being loaded and enabled.

=== Add Nodes to Kubernetes

==== Install Agent

> The distributed component of the Stackable Platform is the agent, which needs to be installed on all managed servers and configured to have access to Kubernetes.

Reviewer 3: "The distributed component" reads weid and doesn't make sense to me.

```suggestion
Tthe agent needs to be installed on all servers that should run services and they need to be configured to have access to Kubernetes.
```
Or similar

> Installing the agent can be done from our rpm and deb package repositories, which we enabled when preparing the nodes above.

Reviewer 3: ```suggestion
Installing the agent can be done from our RPM and deb package repositories, which we enabled when preparing the nodes above.
```

The packages also install a systemd service to run the agent, this is not enabled or started by default though

===== RHEL

    *coming soon*

===== Ubuntu

    apt-get install stackable-agent

==== Configuring the Agent
When installing from packages the agent config file is created in '/etc/stackable-agent/agent.conf'.

The Agent takes a few configuration options, please refer to https://github.com/stackabletech/agent#command-line-parameters[the agent documentation] for more details.

> If you have provided a KUBECONFIG for the root user in `/root/.kube/config` then the agent will use this, if you decided against a _global_ KUBECONFIG, you will need to specify where the config should be read from.

Reviewer 3: Again consistency, KUBECONFIG vs KubeConfig vs Kubeconfig

Author: KUBECONFIG in this case refers to an environment variable, which by convention should be spelled in all caps I believe. I've fixed the other occurrences to all lower case which seems to be consistend with the kubernetes config.
https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/

This can be done by adding a systemd drop-in file that is merged with the package provided unit file.
Put the following content in `/usr/lib/systemd/system/stackable-agent.service.d/kubeconfig.conf`:

----
Environment="KUBECONFIG=/path/to/kubeconfig"
----

==== Starting the Agent
The agent can be started like any regular systemd service by running

> ` systemctl start stackable-agent`

Reviewer 5: The extra space before systemctl breaks the formatting

To enable it to be started at every boot:

> ` systemctl enable stackable-agent`

Reviewer 5: The extra space before systemctl breaks the formatting

During the first start of the agent, it will perform some bootstrapping tasks, most notably it will generate a keypair and request a signed certificate from Kubernetes.

If your Kubernetes is not configured to auto-approve these CSRs, you will need to manually approve that request before the agent can start.

You can do this by running `kubectl certificate approve <agent-hostname>-tls` on the orchestrator server after starting the agent.

If your Kubernetes does not offer certificate signing services, you can alternatively also manually generate a keypair and sign a certificate with a proper authority.
These files will then need to be configured in the agent config file.

=== Health Check
After performing the steps outlined above you should have an empty Stackable Platform up and running.

> When you run `kubectl nodes get` on the orchestrator, you should see the objects that all your agents added to indicate their readiness.

Reviewer 5: The arguments seem to be permuted: `kubectl get nodes` worked fine.

However, the Stackable Platform only uses a very limited subset of the full Kubernetes features, so for our purposes we could actually get away with just running a https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/[kube-apiserver] and leaving out the entire rest of the https://kubernetes.io/docs/concepts/overview/components/[control-plane].
This makes the deployment a lot simpler.

> NODE: We have tested our stack with Kubernetes 1.18 and above

Reviewer 3: ```suggestion
NOTE: We have tested our stack with Kubernetes 1.18 and above
```

----

> ==== Distribute KubeConfig

Reviewer 3: This is still off?

The Agent takes a few configuration options, please refer to https://github.com/stackabletech/agent#command-line-parameters[the agent documentation] for more details.

> If you have provided a kubeconfig for the root user in `/root/.kube/config` then the agent will use this, if you decided against a _global_ `KUBECONFIG` enwironment variable, you will need to specify where the config should be read from.

Reviewer 3: ```suggestion
If you have provided a kubeconfig for the root user in `/root/.kube/config` then the agent will use this, if you decided against a _global_ `KUBECONFIG` environment variable, you will need to specify where the config should be read from.
```
